src/dragonball/src/dbs_pci/src/msi.rs

```diff
@@ -0,0 +1,759 @@
+                    LittleEndian::write_u16(data, self.mask_bits as u16);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 2 => {
+                    LittleEndian::write_u16(data, (self.mask_bits >> 16) as u16);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 4 => {
+                    self.get_pending_state(data, 0);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 6 => {
+                    self.get_pending_state(data, 2);
+                }
+                _ => debug!("invalid offset"),
+            },
+            4 => match offset {
+                0x0 => LittleEndian::write_u32(
+                    data,
+                    self.cap_id_next as u32 | ((self.msg_ctl as u32) << 16),
+                ),
+                MSI_MSG_ADDR_LO_OFFSET => LittleEndian::write_u32(data, self.msg_addr_lo),
+                x if addr_hi_offset.is_some() && x == addr_hi_offset.unwrap() => {
+                    LittleEndian::write_u32(data, self.msg_addr_hi);
+                }
+                x if x == msg_data_offset => {
+                    LittleEndian::write_u32(data, self.msg_data as u32);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                    LittleEndian::write_u32(data, self.mask_bits);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 4 => {
+                    self.get_pending_state(data, 0);
+                }
+                _ => debug!("invalid offset"),
+            },
+            _ => debug!("invalid data length"),
+        }
+
+        Ok(())
+    }
+
+    /// Handle write accesses to the PCI MSI capability structure.
+    pub fn write(&mut self, offset: u64, data: &[u8]) -> std::io::Result<()> {
+        let (msg_data_offset, addr_hi_offset, mask_bits_offset) = self.get_offset_info();
+
+        // Update cache without overriding the read-only bits.
+        match data.len() {
+            1 => match offset {
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                    self.mask_bits = data[0] as u32 | (self.mask_bits & 0xffff_ff00);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 1 => {
+                    self.mask_bits = (data[0] as u32) << 8 | (self.mask_bits & 0xffff_00ff);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 2 => {
+                    self.mask_bits = (data[0] as u32) << 16 | (self.mask_bits & 0xff00_ffff);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 3 => {
+                    self.mask_bits = (data[0] as u32) << 24 | (self.mask_bits & 0x00ff_ffff);
+                }
```

Comment:
Please give some comments on each arm.

Reply:
this is for writing bits according to pci spec. I have added a summarised comment above
"Support writing mask_bits, MSI control information, message data, message address."